Add v2 endpoints

emerkle826 opened the pull request:
Fixes #326 

This PR adds/updates v2 endpoints endpoint (I know) to support new operations to be used by Cassandra Reaper

DO NOT MERGE this until #324 is approved and merge. The rebase this branch off that merge into master

adejanovski commented:
@emerkle826, I took the liberty of moving this PR to draft mode to make it clear it shouldn't be merged. You can take it out of draft mode when the time comes.

Miles-Garnsey reviewed (commented):
Sorry, I'll need a bit more time on this one. I left a couple of questions :)

emerkle826 commented on management-api-server/src/test/resources/com/datastax/mgmtapi/operator-sample.json:
```diff
@@ -0,0 +1,78 @@
+{
```
Good question. There was already a test that was calling `configureNode` that passed yaml [here](https://github.com/k8ssandra/management-api-for-apache-cassandra/blob/master/management-api-server/src/test/java/com/datastax/mgmtapi/LifecycleIT.java#L141-L159). However, the API actually takes JSON or YAML as the request body (JSON [here](https://github.com/k8ssandra/management-api-for-apache-cassandra/blob/master/management-api-server/src/main/java/com/datastax/mgmtapi/resources/LifecycleResources.java#L325-L342), YAML [here](https://github.com/k8ssandra/management-api-for-apache-cassandra/blob/master/management-api-server/src/main/java/com/datastax/mgmtapi/resources/LifecycleResources.java#L344-L358)). If you look, you'll see the method that handles JSON essentially tries to read the Request bosy as a JSON string, then convert that JSON into YAML, then just call the `configureNode` method that handles YAML and does the real work.

Having 2 methods both handling the same API endpoint (POST to `configure` effectively) causes the OpenAPI spec generation to fail (it actually only documents one of the methods, so the doc only shows JSON or only shows YAML as the Request body type, I can't remember which). That is what led to the wonderful hack of marking both those methods as `@Hidden` and manually documenting them [here](https://github.com/k8ssandra/management-api-for-apache-cassandra/blob/master/management-api-server/src/main/resources/openapi-configuration.json).

In this PR, I have fixed all of that some. First, I consolidated the OpenAPI annotations for `@Consumes` into a single method that accepts YAML or JSON now:
https://github.com/k8ssandra/management-api-for-apache-cassandra/pull/325/files#diff-f0ea4ea1635e425b16c7279f1a2e1c3f5eec7fd96cb5e6c2c14774af2533519bR315
I changed the implementation of that method a little since it can't assume it's only handling JSON requests. So now it tries to parse the request body into JSON, then tries to convert that to YAML.
https://github.com/k8ssandra/management-api-for-apache-cassandra/pull/325/files#diff-f0ea4ea1635e425b16c7279f1a2e1c3f5eec7fd96cb5e6c2c14774af2533519bR364-R368
If that fails, it might be that the request body was already yaml, so in the catch block, it just passes the request body as a string to the `configureNode` method that deals with YAML:
https://github.com/k8ssandra/management-api-for-apache-cassandra/pull/325/files#diff-f0ea4ea1635e425b16c7279f1a2e1c3f5eec7fd96cb5e6c2c14774af2533519bR369-R371

Looking at it now, I kept the original method names, but I should probably rename them as they don't do what the names indicate.

emerkle826 replied:
Totally forgot to mention, the yaml version of this file already existed in the test resources. I created this json version from that file simply to test that the json endpoint works as well as the yaml endpoint.

Miles-Garnsey replied:
OK, this isn't a bad way to do it, however could we possibly control the expected type using the `Content-Type` header instead of using a try-catch block? That seems like the principled way to do it. Does that make things harder for you with the spec generation, or easier?

emerkle826 replied:
The `@Consumes` annotation effectively defines the `Content-Type`s that are allowed. For the `/configure` endpoint, it is `application/json`, `application/yaml` and `text/yaml`.

Prior to this change, there were 2 methods attached to the `/configure` endpoint; one that took JSON, and one that took YAML.
https://github.com/k8ssandra/management-api-for-apache-cassandra/blob/master/management-api-server/src/main/java/com/datastax/mgmtapi/resources/LifecycleResources.java#L326
```java
  @POST
  @Consumes("application/json")
  @Operation(
      description = "Configure Cassandra. Will fail if Cassandra is already started",
      operationId = "configureNodeJson")
  @Produces(MediaType.TEXT_PLAIN)
  public synchronized Response configureNodeJson(
      @QueryParam("profile") String profile, String config) {
```

https://github.com/k8ssandra/management-api-for-apache-cassandra/blob/master/management-api-server/src/main/java/com/datastax/mgmtapi/resources/LifecycleResources.java#L348

```java
  @POST
  @Consumes({"application/yaml", "text/yaml"})
  @Operation(
      description = "Configure Cassandra/DSE. Will fail if Cassandra/DSE is already started",
      operationId = "configureNode")
  @Produces(MediaType.TEXT_PLAIN)
  public synchronized Response configureNode(@QueryParam("profile") String profile, String yaml) {
```
However, the OpenAPI document generation did not like 2 methods associated with the same (POST) operation on the same path, with  the Content-Type of the request being the only difference. Initially, this led me to excluding both methods from the auto-generated OpenAPI doc and maintaining the definition it should look like in that static file (the one that gets merged with the auto-generated stuff to produce the final OpenAPI document).

This PR tries to reduce the manual maintenance required to keep the doc accurate. To do so, I have to consolidate the POST to `/configure` into a single method so the auto-generation doesn't choke, which means one method that accepts both Content-Types.
```java
  public synchronized Response configureNodeJson(
      @QueryParam("profile") String profile,
      @RequestBody(
              content = {
                @Content(mediaType = "application/yaml"),
                @Content(mediaType = "text/yaml"),
                @Content(mediaType = "application/json")
              },
              required = true,
              description = "JSON or YAML configuration overrides")
          String config) {
```
Unfortunately, doing that means I can no longer tell the request body format as now I have a String representation of either JSON or YAML configuration. If you look at the old implementation, the method that handled JSON requests just converted the JSON to YAML and then called the method that handled YAML requests, which did all the work. The benefit however is that if parsing the request body as JSON and then converting to YAML failed, we could return an error saying "you have bad JSON".

With my changes, bad JSON will just get attempted again as YAML, and likely still be bad, which will return an error indicating a bad request. It's not as specific, but the behavior for the success paths is exactly the same.

There is likely a better way of doing this, but I don't want to change the `v0/configure` endpoint and how it currently works. We can address it in v2 (maybe we just expose 2 endpoints to avoid the doc generation and mixing requestBody Content-Type handling into a single method) if we want, but it's out of scope for the tickets in this phase.

Lastly, if this change, which arguably is out of scope as well, is too much for this PR, I can revert it, but it will bring back the hand-managed definition for the `confugre` endpoint.

Miles-Garnsey replied:
Per discussions with @emerkle826 we are deferring this work until later. We will stick with two endpoints for now and manually generate the OpenAPI doc.

emerkle826 commented on management-api-server/doc/openapi.json:
```diff
@@ -7,12 +7,16 @@
       "url" : "http://www.apache.org/licenses/LICENSE-2.0.html"
```
Another good question. It's a little of both.

First, the [plugin](https://github.com/k8ssandra/management-api-for-apache-cassandra/blob/master/management-api-server/pom.xml#L291-L312) is configured to run at compile time during the resolve phase so that all of the _server_ module code is compiled before trying to make sense of the annotations. The annotations on each of the methods in the server maven sub-module (the one that defines all the endpoints) drives the bulk of the doc generation. However, the jaxrs Application setup in [ManagementApplication](https://github.com/k8ssandra/management-api-for-apache-cassandra/blob/master/management-api-server/src/main/java/com/datastax/mgmtapi/ManagementApplication.java#L28-L76) doesn't quite setup everything needed for the doc generation to happen via annotations alone. So, the [plugin](https://github.com/k8ssandra/management-api-for-apache-cassandra/blob/master/management-api-server/pom.xml#L300) is also configured to merge the auto-generated doc with a static file that provides the rest of the bits [here](https://github.com/k8ssandra/management-api-for-apache-cassandra/blob/master/management-api-server/src/main/resources/openapi-configuration.json).

Ideally we would have the doc completely auto-generated, but that is a little out of scope here. I did manage to remove the configure endpoint from the static file, so there's even less that is statically managed.

emerkle826 replied:
actually, this whole doc change I think needs to be reverted and moved to the other PR. Some of the changes in here reflect work done in the other PR.....

Miles-Garnsey replied:
Cool, so basically we have a bit of dodginess around the configuration endpoint and we are working around that by defining it manually, then everything else gets merged in? I can live with that for now :)

emerkle826 replied:
> Cool, so basically we have a bit of dodginess around the configuration endpoint and we are working around that by defining it manually, then everything else gets merged in? I can live with that for now :)

I guess I should have read this first :smile: If this is what you would prefer, then the whole discussion above about Content-Type separation can go away for now, and I can revert that bit.

Miles-Garnsey replied:
Let's do a quick catch up tomorrow if possible just to close this out. If you have a solution that you're happier with then I think keep it, but I just want to make sure I understand the implications of both solutions.

Miles-Garnsey replied:
From discussions with @emerkle826 we are going to defer re-working this for later. For now we will continue to manually generate the OpenAPI stuff associated with these endpoints rather than combining them into one endpoint.

emerkle826 commented on management-api-server/src/main/resources/openapi-configuration.json:
```diff
@@ -2,89 +2,20 @@
     "openAPI": {
         "info": {
-            "version": "0.1",
+            "version": "0.1.66",
```
This will be where the verssion change to 0.1.67 has to happen (removing the other comment)

Miles-Garnsey commented:
From discussions earlier today, if we already have a test framework for the management API, could we please get some e2e tests on this so that all of the new client methods are exercised - once that's done, I think I'll be pretty confident in this PR :) 

The other alternative is to check the tests that are added as part of the code gen - @emerkle826 if you are comfortable that those are sufficient we can stick with them. Bearing in mind that this will all be e2e tested in Reaper's tests anyway (but it is still nice to have thorough tests here too...)

emerkle826 commented:
> From discussions earlier today, if we already have a test framework for the management API, could we please get some e2e tests on this so that all of the new client methods are exercised

Testing the generated client methods will be difficult with how the generation happens currently. The client source and tests are generated with every compile, but they are not compiled during the compile of the project. Compiling the client requires you to issue another Maven command (i.e. `mvn package` or `mvn install`) in the generated project folder, or by providing the generated project's pom.xml location:

```sh
mvn install -f ./management-api-server/target/generated-sources/openapi/java-client/pom.xml
```
from the root of this project.

This is a bit of a chicken-and-egg problem in that generated sources are excluded from git version control (things like build artifacts and stuff that is compiled every time you build). So there is no generated client that is actually checked in to git source control.

The other issue for the generated client is that tests are also generated, but they are practically useless. All of the tests generated are effectively unit tests, and none of them do any validation other than that the client methods can be called, which is really just a validation of the compile-ability of the generated source.

emerkle826 commented:
> The other alternative is to check the tests that are added as part of the code gen - @emerkle826 if you are comfortable that those are sufficient we can stick with them.

Unfortunately, they are useless.

emerkle826 commented:
@Miles-Garnsey I've reworked this PR/branch. I've pulled out some of the changes that are not scoped specifically to implementing the v2 endpoints endpoint. I've reverted the `configure` endpoint changes for now as well.

This PR still should be rebased after PRs #333, #334 and #335 are merged.

emerkle826 commented:
@Miles-Garnsey Gonna shelve this version for now in favor of #343. I will revisit this PR for a refactor week. The current endpoint returns what we need and #343 fixes the annotations so that the generated client has the proper return type.

emerkle826 commented:
Closing this as it will be handled in a refactor week
